From e662165793c3d6ef65d4ac6d465e0fa6c38b98d2 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Fri, 5 Dec 2025 15:50:21 +0100
Subject: [PATCH 945/954] feat(stm32mp1): implement a wake up of the primary
 CPU in stm32_cpu_standby

Let the secondary CPU wake up the primary CPU by raising an SGI.
This SGI will be cleared in OPTEE.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git a2c47f42b987e73f85c1ae193d5a1650fe9b2899 feat(stm32mp1): implement a wake up of the primary CPU in stm32_cpu_standby]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I42aa2db747487f515a0bd65b86d6dbf713f81fe1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/515705
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp1/bl2_plat_setup.c        | 4 ++++
 plat/st/stm32mp1/platform.mk             | 6 ++++++
 plat/st/stm32mp1/stm32mp1_critic_power.c | 6 ++++++
 3 files changed, 16 insertions(+)

diff --git a/plat/st/stm32mp1/bl2_plat_setup.c b/plat/st/stm32mp1/bl2_plat_setup.c
index 6978391cdb..aa107b57b4 100644
--- a/plat/st/stm32mp1/bl2_plat_setup.c
+++ b/plat/st/stm32mp1/bl2_plat_setup.c
@@ -499,6 +499,10 @@ skip_console_init:
 
 	print_reset_reason();
 
+#if !STM32MP1_OPTEE_IN_SYSRAM
+	stm32mp_gic_init();
+#endif
+
 #if STM32MP15
 	if (stm32mp_check_closed_device() == STM32MP_CHIP_SEC_CLOSED) {
 		update_monotonic_counter();
diff --git a/plat/st/stm32mp1/platform.mk b/plat/st/stm32mp1/platform.mk
index d9c4c7c490..e14809dbc1 100644
--- a/plat/st/stm32mp1/platform.mk
+++ b/plat/st/stm32mp1/platform.mk
@@ -137,6 +137,12 @@ $(error Cannot find $(patsubst %.dtb,%.dts,$(DTB_FILE_NAME)) file)
 endif
 endif
 
+# Generic GIC v2
+include drivers/arm/gic/v2/gicv2.mk
+BL2_SOURCES		+=	${GICV2_SOURCES}
+BL2_SOURCES		+=	plat/common/plat_gicv2.c
+BL2_SOURCES		+=	plat/st/common/stm32mp_gic.c
+
 # Macros and rules to build TF binary
 STM32_TF_STM32		:=	$(addprefix ${BUILD_PLAT}/tf-a-, $(patsubst %.dtb,%.stm32,$(DTB_FILE_NAME)))
 STM32_LD_FILE		:=	plat/st/stm32mp1/stm32mp1.ld.S
diff --git a/plat/st/stm32mp1/stm32mp1_critic_power.c b/plat/st/stm32mp1/stm32mp1_critic_power.c
index 2621d15695..5f2f3d8937 100644
--- a/plat/st/stm32mp1/stm32mp1_critic_power.c
+++ b/plat/st/stm32mp1/stm32mp1_critic_power.c
@@ -6,6 +6,7 @@
 
 #include <arch_helpers.h>
 #include <common/debug.h>
+#include <drivers/arm/gic_common.h>
 #include <drivers/arm/gicv2.h>
 #include <drivers/generic_delay_timer.h>
 #include <drivers/st/stm32_iwdg.h>
@@ -113,6 +114,7 @@ static void stm32mp1_power_unlock(void) {
 
 static void stm32_cpu_standby(void) {
 	unsigned int pos = plat_my_core_pos();
+	unsigned int other_pos = stm32_get_pos_other();
 
 	stm32mp1_power_lock();
 	cstop_cpux_state[pos] = STATE_AUTO_CSTOP_ENTRY;
@@ -131,6 +133,10 @@ static void stm32_cpu_standby(void) {
 
 	stm32mp1_power_lock();
 
+	/* Use SGI_6 to wake up the other core, the ack will be done in OP-TEE */
+	if (cstop_enter == STATE_AUTO_CSTOP_ENTRY)
+		gicv2_raise_sgi(ARM_IRQ_SEC_SGI_6, false, other_pos);
+
 	cstop_cpux_state[pos] = STATE_NONE;
 
 	stm32mp1_power_unlock();
-- 
2.53.0

